Add support for Icarus keys derived from 15-word mnemonic sentences

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1208 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- commit 798137c6893bce147132dc2b4208b501872b4147

    enable 15-word mnemonic sentence in byron wallet creation
    
    This, to effectively allow restoring legacy Yoroi wallets through the same endpoint'

- commit 6e14ff15f4ae293a5021f0c019fd537b0235786f 

    start setting up server with Icarus key to illustrate how it would look like

- commit 2fcd0db628ed6f406e99646e835bec12469cefb4 

    Add a new key target 'Icarus' to capture the era in-between Byron and Shelley
    
    Icarus keys are a variant of BIP-44, but working on top of Byron addresses. So they are an
    hybrid form between Byron and Shelley. Note that we only support the 'Mainnet variation of them.
    
- commit 9a20cea7999fd6ec6dd8d0f2f9659304e1221985

    add unit tests for AddressDerivation.Icarus module
    
    - Moved 'genAddress' & 'genLegacyAddress' to AddressDerivationSpec for sharing.
    - Added XPrv and XPub DB roundtrips
    - Added roundtrip for MkFingerprints
    - Resurrected golden specs for Yoroi

- commit 3aaefcb8318bc33f4f6062c04fc5797a8f34a916  and commit cca28956ff727f70647091877eda9ce0a28fc5d9

    reduce duplication in SequentialSpec and make properties polymorphic in keys
    
    - This also makes for a more readable test output
    - Next step: tweaking the generator so that we can have both Icarus and Shelley keys tested through this mean

- commit 0cf8a54aa99597e982ea18de94c084d843628b12

    instantiate an extra Icarus API when starting server

- commit 6d22603ae391767e75c8bfca0aee2918ae2b0138 

    properly discriminate between icarus & byron api layers in API handlers


# Comments

<!-- Additional comments or screenshots to attach if any -->

Note: as an extra test, I also restored a known Byron Yoroi wallet manually on the ITN. Transactions showed up and it seemed fine :tada: 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
enable 15-word mnemonic sentence in byron wallet creation

This, to effectively allow restoring legacy Yoroi wallets through the same endpoint'

KtorZ committed:
Add a new key target 'Icarus' to capture the era in-between Byron and Shelley

Icarus keys are a variant of BIP-44, but working on top of Byron addresses. So they are an
hybrid form between Byron and Shelley. Note that we only support the 'Mainnet variation of them.

TODO:
Port relevant unit tests for testing Icarus key fully, in particular, golden tests with Yoroi.

KtorZ committed:
start setting up server with Icarus key to illustrate how it would look like

KtorZ committed:
add unit tests for AddressDerivation.Icarus module

- Moved 'genAddress' & 'genLegacyAddress' to AddressDerivationSpec for sharing.
- Added XPrv and XPub DB roundtrips
- Added roundtrip for MkFingerprints
- Resurrected golden specs for Yoroi

KtorZ committed:
reduce duplication in SequentialSpec and make properties polymorphic in keys

- This also makes for a more readable test output
- Next step: tweaking the generator so that we can have both Icarus and Shelley keys tested through this mean

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -561,12 +568,22 @@ mkShelleyWallet ctx wid cp meta pending progress = do
+    let ApiMnemonicT (_, mnemonic) = body ^. #mnemonicSentence
+    case length mnemonic of
+        n | n == 12 -> postByronWallet byron body
+        n | n == 15 -> postIcarusWallet icarus body
```
:+1:

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -183,9 +197,16 @@ spec = do
+        it "XPrv IcarusKey"
+            (property $ prop_roundtripXPrv @IcarusKey)
+        it "XPrv ByronKey"
+            (property $ prop_roundtripXPrv @ByronKey)
```
why not `prop_roundtripXPub @ByronKey` ?

KtorZ replied:
Because there's no 'SerializeXPub' instance on Byron keys. We don't need to serialize public keys to binary for these :)

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/SequentialSpec.hs:
```diff
@@ -510,14 +523,21 @@ instance Typeable chain => Arbitrary (AddressPool chain ShelleyKey) where
+-- properties.
+data Style =
+    forall (chain :: AccountingStyle). Typeable chain => Style (Proxy chain)
+instance Show Style where show (Style proxy) = show (typeRep proxy)
```
:+1:

KtorZ committed:
enable address discovery properties for IcarusKey as well

KtorZ committed:
instantiate an extra Icarus API when starting server

KtorZ committed:
properly discriminate between icarus & byron api layers in API handlers

KtorZ committed:
make BYRON integration tests work with both Icarus & Byron

- Added faucet Icarus wallets
- Added 'fixtureIcarusWallet' and 'emptyIcarusWallet' and use it to implement scenarios using both wallets
